### PR TITLE
Refactor frontend to use QML / QtQuick.

### LINKED ISSUE #956: Migrate frontend from Qt Widgets to Qt QML API
We have a PR for that one already, but this is the tracking ticket.

We basically depend on #955 before we can continue working on this, in order to make it way simpler to use QML (as we only need to care about latest major Qt version)

Depends on #955.

### PR DESCRIPTION
**Pull request by christianparpart**
### Motivation

- More future-proof / forward compatible Qt API usage
- Hopefully easier to create new UI elements using QML / QtQuick rather than QtWidgets.
- Basinc new ideas on top of this PR, specifically: Tabs and later on split views.
- Having a unified look'n'feel of UI elements. In fact, the scrollbar currently doesn't look nice on all systems, e.g. @whisperity seems to have the default style while I have the awesome style on 0.3.2. With moving the QML/QtQuick it **seems** possible to have the same look on all platforms and desktops.

This PR "simply" refactors the frontend and ports it from QtWidgets (QOpenGLWidget) to QML / QtQuick. I try not to add any new feature or anything but get it on feature parity.

### Checklist

- [x] port to `QQuickItem` for the Widget
- [x] expose `TerminalSession` to QML
- [x] ensure functional scrollbar
- [x] ensure window title updates
- [x] use `implicitWidth` / `implicitHeight` rather than `width` / `height` to declare widget component size
- [x] system beep working
- [x] BUG: currently app termination is assert()ing
- [x] ensure blurred background is working
- [x] UI dialogs working, such as permission request
- [x] notifications-OSC working (Requires Qt 5.12 or higher)
- [x] IME working
- [x] [PLATFORM] ensure proper rendering on Wayland
- [x] [PLATFORM] Ensure it's running on Windows
- [x] scrollbar (move the current JS impl to C++ by introducing a `ScrollBar.vertical` property (similar to QtQuick's))
  - [x] hide_in_alt_screen:
- [ ] platforms (ox/windows/linux): ensure creating new windows in-process and out-of-process is working
   - [x] Windows 
   - [x] Linux
- [x] resize
- [x] background image
  - [x] background is loaded from the path
  - [x] opacity of the background image
  - [x] proper layer positions 
- [x] opacity 
- [x] clean up
Closes #956.

### CODE REVIEW

**Comment by uspasojevic96** [comment]
Also add to TODO to not break the functionality on macOS where a single process can have multiple windows instead of spawning entirely new process

File: src/qml_generator/generator.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        data["import"]["QtQuicklabs"] = "Qt.labs.platform \n";
+        data["import"]["QtMultimedia"] = "QtMultimedia  \n";
+        data["import"]["QtGraphicalEffects"] = "QtGraphicalEffects \n";
+
+    };
+    inja::json data;
+    #ifdef GENERATE_QT5
+    filldataQt5(data);
+    #else
+    filldataQt6(data);
```
Comment (christianparpart) [inline src/qml_generator/generator.cpp:60]: sounds like what you want (to stay in line with the other helper methods), is something like `createModuleNames(data)`, a single global function, and then have the `#ifdef GENERATE_QT5` inside of this body.

All of the above functions are private, and can thus be moved inside an anonymous namespace to hide their symbols and mark them as TU-private (`namespace { ... }`)

**Comment by Utkarsh-khambra** [comment]
I get following errors when I run contour built from this branch
```
Critical:
ERROR: No native SystemTrayIcon implementation available.
Qt Labs Platform requires Qt Widgets on this setup.
Add 'QT += widgets' to .pro and create QApplication in main().
 ((null):0, (null))
Critical:
ERROR: No native Menu implementation available.
Qt Labs Platform requires Qt Widgets on this setup.
Add 'QT += widgets' to .pro and create QApplication in main().
 ((null):0, (null))
Warning: QSoundEffect(qaudio): Error decoding source qrc:/contour/bell.oga ((null):0, (null))
[error] OpenGL error 1282 for call: glTexImage2D(target, levelOfDetail, (int) glFmt, unbox<int>(param.size.width), unbox<int>(param.size.height), UnusedParam, glFmt, type, stub.data())
[error] OpenGL error 1282 for call: glPixelStorei(GL_UNPACK_ALIGNMENT, param.rowAlignment)
[error] OpenGL error 1282 for call: glPixelStorei(GL_UNPACK_ALIGNMENT, param.rowAlignment)
[error] OpenGL error 1282 for call: glPixelStorei(GL_UNPACK_ALIGNMENT, param.rowAlignment)
[error] OpenGL error 1282 for call: glPixelStorei(GL_UNPACK_ALIGNMENT, param.rowAlignment)
h[error] OpenGL error 1282 for call: glPixelStorei(GL_UNPACK_ALIGNMENT, param.rowAlignment)
[error] OpenGL error 1282 for call: glPixelStorei(GL_UNPACK_ALIGNMENT, param.rowAlignment)
[error] OpenGL error 1282 for call: glPixelStorei(GL_UNPACK_ALIGNMENT, param.rowAlignment)
[error] OpenGL error 1282 for call: glPixelStorei(GL_UNPACK_ALIGNMENT, param.rowAlignment)
[error] OpenGL error 1282 for call: glPixelStorei(GL_UNPACK_ALIGNMENT, param.rowAlignment)
[error] OpenGL error 1282 for call: glDeleteVertexArrays(1, &_rectVAO)
 
```
Though contour seems to work fine. I am running contour on Fedora 37 with Gnome 43.5

**Comment by Yaraslaut** [comment]
Crash that i see on qml branch 
```
[/home/yaraslau/repo/contour/src/vtrasterizer/TextRenderer.cpp:855] Precondition failed. rowCount <= unbox<unsigned>(glyph.bitmapSize.height)
==20042==
==20042== Process terminating with default action of signal 6 (SIGABRT): dumping core
==20042==    at 0x75E526C: __pthread_kill_implementation (pthread_kill.c:44)
==20042==    by 0x7595A07: raise (raise.c:26)
==20042==    by 0x757E5C1: abort (abort.c:100)
==20042==    by 0x2F9759: crispy::detail::fail(std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, int) (assert.h:84)
==20042==    by 0x448B42: terminal::rasterizer::TextRenderer::createRasterizedGlyph(terminal::rasterizer::atlas::TileLocation, text::glyph_key const&, unicode::PresentationStyle) (TextRenderer.cpp:855)
==20042==    by 0x449223: terminal::rasterizer::TextRenderer::createSlicedRasterizedGlyph(terminal::rasterizer::atlas::TileLocation, text::glyph_key const&, unicode::PresentationStyle, crispy::StrongHash const&) (TextRenderer.cpp:724)
==20042==    by 0x449D22: operator() (TextRenderer.cpp:712)
==20042==    by 0x449D22: constructTile<terminal::rasterizer::TextRenderer::getOrCreateRasterizedMetadata(const crispy::StrongHash&, const text::glyph_key&, unicode::PresentationStyle)::<lambda(terminal::rasterizer::atlas::TileLocation)> > (TextureAtlas.h:563)
```

**Comment by christianparpart** [comment]
> Crash that i see on qml branch
> 
> ```
> [/home/yaraslau/repo/contour/src/vtrasterizer/TextRenderer.cpp:855] Precondition failed. rowCount <= unbox<unsigned>(glyph.bitmapSize.height)
> ==20042==
> ==20042== Process terminating with default action of signal 6 (SIGABRT): dumping core
> ==20042==    at 0x75E526C: __pthread_kill_implementation (pthread_kill.c:44)
> ==20042==    by 0x7595A07: raise (raise.c:26)
> ==20042==    by 0x757E5C1: abort (abort.c:100)
> ==20042==    by 0x2F9759: crispy::detail::fail(std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >, int) (assert.h:84)
> ==20042==    by 0x448B42: terminal::rasterizer::TextRenderer::createRasterizedGlyph(terminal::rasterizer::atlas::TileLocation, text::glyph_key const&, unicode::PresentationStyle) (TextRenderer.cpp:855)
> ==20042==    by 0x449223: terminal::rasterizer::TextRenderer::createSlicedRasterizedGlyph(terminal::rasterizer::atlas::TileLocation, text::glyph_key const&, unicode::PresentationStyle, crispy::StrongHash const&) (TextRenderer.cpp:724)
> ==20042==    by 0x449D22: operator() (TextRenderer.cpp:712)
> ==20042==    by 0x449D22: constructTile<terminal::rasterizer::TextRenderer::getOrCreateRasterizedMetadata(const crispy::StrongHash&, const text::glyph_key&, unicode::PresentationStyle)::<lambda(terminal::rasterizer::atlas::TileLocation)> > (TextureAtlas.h:563)
> ```

This crash certainly is not related to this PR. It looks like it most definitely will also raise that assert() outside this PR. 

EDIT: Okay, looks like you fixed that one already on master, right? Well yeah. ^^

File: src/contour/TerminalSession.cpp [hunk 2]
```diff
@@ -1128,14 +1234,15 @@ void TerminalSession::spawnNewTerminal(string const& _profileName)
 
     if (config_.spawnNewProcess)
     {
+        printf("spawning new process\n");
```
Comment (Yaraslaut) [inline src/contour/TerminalSession.cpp:1237]: Is it correct? shouldn't it be some log? And same for line 1243

File: src/contour/display/OpenGLRenderer.cpp [hunk 3]
```diff
@@ -861,6 +975,8 @@ void OpenGLRenderer::initializeBackgroundRendering()
 
 void OpenGLRenderer::setBackgroundImage(shared_ptr<terminal::BackgroundImage const> const& backgroundImageOpt)
 {
+    return;
+    // TODO(pr): this crashes on Qt6+QtQuick due to too early use of OpenGL during init
```
Comment (Yaraslaut) [inline src/contour/display/OpenGLRenderer.cpp:979]: make PR

**Pull request merged**